### PR TITLE
[BATCHED_RPA] Add SEQ_ON_LANE kv cache layout to remove padding when kv_heads = 1, kv d_type = fp8

### PR DESCRIPTION
**Pull request by a1yssan13**
# Description

In RPA, the kv_cache layout is packed along the number of KV heads. Thus, when `num_kv_heads = 1`, and kv dtype = fp8, we have to waste half the KV cache with padding. (k + v = 2, but 2 < kv_packing = 4). 

We allow the option to keep the cache in a layout -- seq_along_lane, with minimal performance degradation. In this layout, we put page_size on the last dimension (lane), and the head dimension on the sublane dimension.

Feature in progress to remove any overhead introduced by using `SEQ_ON_LANE` 

You can enable this via flag: `USE_BATCHED_RPA_SEQ_ON_LANE=1`

# Features 

With `SEQ_ALONG_LANE` enabled: 

1. You can now use `batched_rpa` with models that have`head_dim = 64`, since we no longer have to pad the head dimension up to 128. To be specific, SEQ_ALONG_LANE relaxes the constraint of having a head dimension that must be a multiple of 128 to only a multiple of 32. 
2. KV cache utilization is halved in models with the above characteristics (kv_heads per device = 1, kv_packing = 4)

# Constraints 

Since PAGE_SIZE is now on the last dimension of our cache, the configured block size must be 128 (for now). 

# Tests

All tests are run on Qwen-Coder. But any model that uses `BATCHED_RPA` should benefit. 

KV Cache capacity is doubled: 

`USE_BATCHED_RPA_SEQ_ON_LANE=0`:
```text
(EngineCore pid=2929407)   0.1.0: Array(shape=(5141, 128, 8, 4, 128), dtype=float8_e4m3fn, sharding=NamedSharding(mesh=Mesh('data': 1, 'model': 8, axis_types=(Auto, Auto)), spec=P(None, None, 'model'), memory_kind=device) [NamedSharding, id=123473401299648])
```

`USE_BATCHED_RPA_SEQ_ON_LANE=1`:
```text
(EngineCore pid=3053252)   0.1.0: Array(shape=(10283, 2, 256, 4, 128), dtype=float8_e4m3fn, sharding=NamedSharding(mesh=Mesh('data': 1, 'model': 8, axis_types=(Auto, Auto)), spec=P(None, None, 'model'), memory_kind=device) [NamedSharding, id=128670224217152])
```

SEQ_ON_LANE adds some overhead at low concurrency, but this layout unlocks double the physical KV cache capacity, allowing higher concurrency / throughput. 

### `1024_1024` @ Concurrency 64 vs 512

| Metric | c=64 (Before) | c=64 (After) | Delta % (c=64) | c=512 (Before) | c=512 (After) | Delta % (c=512) |
| :--- | :---: | :---: | :---: | :---: | :---: | :---: |
| **Max HBM Pages** | 5,141 pages | 10,283 pages | **+100.0%** | 5,141 pages | 10,283 pages | **+100.0%** |
| **Req Throughput (req/s)** | 2.21 | 2.15 | -2.7% | 4.70 | 4.93 | **+4.9%** |
| **Out Throughput (tok/s)** | 2,044.11 | 1,984.88 | -2.9% | 4,328.69 | 4,535.48 | **+4.8%** |
| **Total Throughput (tok/s)** | 4,081.76 | 3,963.48 | -2.9% | 8,672.27 | 9,086.55 | **+4.8%** |
| **Mean TTFT (ms)** | 553.46 | 550.99 | -0.4% | 12,735.95 | 8,876.18 | **-30.3%** |
| **Median TTFT (ms)** | 134.04 | 133.62 | -0.3% | 12,142.22 | 6,099.26 | **-49.8%**|
| **Mean TPOT (ms)** | 29.36 | 30.29 | +3.1% | 90.20 | 86.11 | **-4.5%** |

###  `8192_1024` @ Concurrency 64 vs 128

| Metric | c=64 (Before) | c=64 (After) | Delta % (c=64) | c=128 (Before) | c=128 (After) | Delta % (c=128) |
| :--- | :---: | :---: | :---: | :---: | :---: | :---: |
| **Max HBM Pages** | 5,141 pages | 10,283 pages | **+100.0%** | 5,141 pages | 10,283 pages | **+100.0%** |
| **Req Throughput (req/s)** | 1.17 | 1.14 | -2.5% | 1.14 | 1.33 | **+16.7%** |
| **Out Throughput (tok/s)** | 1,083.89 | 1,053.41 | -2.8% | 1,054.81 | 1,228.44 | **+16.5%** |
| **Total Throughput (tok/s)** | 9,692.46 | 9,419.93 | -2.8% | 9,432.43 | 10,985.13 | **+16.5%** |
| **Mean TTFT (ms)** | 3,396.50 | 3,375.75 | -0.6% | 37,775.69 | 11,604.02 | **-69.3%**|
| **Median TTFT (ms)** | 934.04 | 933.62 | -0.0% | 27,715.05 | 1,431.48 | **-94.8%**|
| **Mean TPOT (ms)** | 53.53 | 55.25 | +3.2% | 71.09 | 84.59 | +19.0% |

Specific Results: 

#### `1024 Input / 1024 Output `

Before: 
```text
============ Serving Benchmark Result ============
Successful requests:                     320       
Benchmark duration (s):                  144.60    
Total input tokens:                      294646    
Total generated tokens:                  295581    
Request throughput (req/s):              2.21      
Output token throughput (tok/s):         2044.11   
Total Token throughput (tok/s):          4081.76   
---------------Time to First Token----------------
Mean TTFT (ms):                          553.46    
Median TTFT (ms):                        176.17    
P99 TTFT (ms):                           3311.63   
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          29.36     
Median TPOT (ms):                        29.60     
P99 TPOT (ms):                           31.61     
---------------Inter-token Latency----------------
Mean ITL (ms):                           29.38     
Median ITL (ms):                         25.77     
P99 ITL (ms):                            119.51    
----------------End-to-end Latency----------------
Mean E2EL (ms):                          27662.70  
Median E2EL (ms):                        27490.51  
P99 E2EL (ms):                           33041.31  
==================================================
```
After: 
```text
============ Serving Benchmark Result ============
Successful requests:                     320       
Benchmark duration (s):                  148.92    
Total input tokens:                      294646    
Total generated tokens:                  295581    
Request throughput (req/s):              2.15      
Output token throughput (tok/s):         1984.88   
Total Token throughput (tok/s):          3963.48   
---------------Time to First Token----------------
Mean TTFT (ms):                          550.99    
Median TTFT (ms):                        177.89    
P99 TTFT (ms):                           3315.58   
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          30.29     
Median TPOT (ms):                        30.51     
P99 TPOT (ms):                           32.56     
---------------Inter-token Latency----------------
Mean ITL (ms):                           30.31     
Median ITL (ms):                         26.70     
P99 ITL (ms):                            119.64    
----------------End-to-end Latency----------------
Mean E2EL (ms):                          28514.83  
Median E2EL (ms):                        28339.38  
P99 E2EL (ms):                           34189.75  
==================================================
```

#### `8192 Input / 1024 Output`
Before:
```text
============ Serving Benchmark Result ============
Successful requests:                     320       
Benchmark duration (s):                  273.22    
Total input tokens:                      2352062   
Total generated tokens:                  296143    
Request throughput (req/s):              1.17      
Output token throughput (tok/s):         1083.89   
Total Token throughput (tok/s):          9692.46   
---------------Time to First Token----------------
Mean TTFT (ms):                          3396.50   
Median TTFT (ms):                        934.04    
P99 TTFT (ms):                           25554.56  
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          53.53     
Median TPOT (ms):                        55.90     
P99 TPOT (ms):                           69.81     
---------------Inter-token Latency----------------
Mean ITL (ms):                           53.71     
Median ITL (ms):                         27.89     
P99 ITL (ms):                            468.04    
----------------End-to-end Latency----------------
Mean E2EL (ms):                          53049.98  
Median E2EL (ms):                        52897.64  
P99 E2EL (ms):                           79638.25  
==================================================
```
After: 
```text
============ Serving Benchmark Result ============
Successful requests:                     320       
Benchmark duration (s):                  281.13    
Total input tokens:                      2352062   
Total generated tokens:                  296143    
Request throughput (req/s):              1.14      
Output token throughput (tok/s):         1053.41   
Total Token throughput (tok/s):          9419.93   
---------------Time to First Token----------------
Mean TTFT (ms):                          3375.75   
Median TTFT (ms):                        933.62    
P99 TTFT (ms):                           25479.34  
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          55.25     
Median TPOT (ms):                        57.41     
P99 TPOT (ms):                           72.21     
---------------Inter-token Latency----------------
Mean ITL (ms):                           55.43     
Median ITL (ms):                         29.67     
P99 ITL (ms):                            469.01    
----------------End-to-end Latency----------------
Mean E2EL (ms):                          54622.23  
Median E2EL (ms):                        54366.01  
P99 E2EL (ms):                           82103.39  
==================================================
```

#### `1024 Input / 8192 Output`
Before:
```text
============ Serving Benchmark Result ============
Successful requests:                     320       
Benchmark duration (s):                  1054.34   
Total input tokens:                      294646    
Total generated tokens:                  2373111   
Request throughput (req/s):              0.30      
Output token throughput (tok/s):         2250.80   
Total Token throughput (tok/s):          2530.25   
---------------Time to First Token----------------
Mean TTFT (ms):                          504.15    
Median TTFT (ms):                        133.44    
P99 TTFT (ms):                           3373.71   
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          27.12     
Median TPOT (ms):                        27.22     
P99 TPOT (ms):                           27.55     
---------------Inter-token Latency----------------
Mean ITL (ms):                           27.13     
Median ITL (ms):                         26.65     
P99 ITL (ms):                            29.85     
----------------End-to-end Latency----------------
Mean E2EL (ms):                          201638.65 
Median E2EL (ms):                        202321.14 
P99 E2EL (ms):                           224372.04 
==================================================
```
After:
```text
============ Serving Benchmark Result ============
Successful requests:                     320       
Benchmark duration (s):                  1107.31   
Total input tokens:                      294646    
Total generated tokens:                  2373111   
Request throughput (req/s):              0.29      
Output token throughput (tok/s):         2143.13   
Total Token throughput (tok/s):          2409.22   
---------------Time to First Token----------------
Mean TTFT (ms):                          516.61    
Median TTFT (ms):                        138.87    
P99 TTFT (ms):                           3250.53   
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          28.52     
Median TPOT (ms):                        28.50     
P99 TPOT (ms):                           29.21     
---------------Inter-token Latency----------------
Mean ITL (ms):                           28.53     
Median ITL (ms):                         28.15     
P99 ITL (ms):                            31.48     
----------------End-to-end Latency----------------
Mean E2EL (ms):                          212048.70 
Median E2EL (ms):                        212200.40 
P99 E2EL (ms):                           235744.64 
==================================================
```

# Checklist

### CODE REVIEW

File: tpu_inference/kernels/experimental/batched_rpa/wrapper.py [hunk 1]
```diff
@@ -362,49 +409,56 @@ def ragged_paged_attention(
...
-        out_dtype: Dtype of output. Defaults to dtype of queries.
-        use_causal_mask: Not used.
-
-    Returns:
-        out: [max_num_tokens, num_q_heads, head_dim]. Output of self attention.
-        new_kv_cache: [num_pages, page_size, cdiv(num_kv_heads * 2, kv_packing),
-            kv_packing, head_dim]. Result of new kv cache where k & vs are
-            concatenated along num kv heads dim.
-    """
+  Args:
```
Comment (kyuyeunk) [inline tpu_inference/kernels/experimental/batched_rpa/wrapper.py]: can you fix the indentation? i.e., maintain 4 spaces?

File: tpu_inference/kernels/experimental/batched_rpa/kernel.py [hunk 2]
```diff
@@ -365,31 +436,31 @@ def rpa_kernel(
 ) -> tuple[jax.Array, jax.Array]:
     """Perform batched ragged paged attention with scheduler data.
 
-    Args:
```
Comment (kyuyeunk) [inline tpu_inference/kernels/experimental/batched_rpa/kernel.py:368]: ditto.

**Review by kyuyeunk — COMMENTED** [review]
left nit comment. but let me try it on e2e first before approval.

**Comment by kyuyeunk** [comment]
can you fix this logic as well? https://github.com/vllm-project/tpu-inference/blob/6a9c528f4d5cdb3d11e90402021b1977ed4aa0ef/tpu_inference/layers/common/sharding.py#L254-L261

**Review by kyuyeunk — APPROVED** [review]

File: tpu_inference/kernels/experimental/batched_rpa/flash_attention.py [hunk 3]
```diff
@@ -44,13 +43,23 @@ def flash_attention_qk_softmax(
...
+            q.reshape(-1, tq, h_size),
+            k.reshape(-1, h_size, s),
+            dimension_numbers=(([2], [1]), ([0], [0])),
+            preferred_element_type=jnp.float32,
+        )
+    else:
+        s = k.shape[-2]
+        qk = lax.dot(
+            q.reshape(b, tq, h_size),
+            k.reshape(b, s, h_size),
```
Comment (lxhfirenking) [inline tpu_inference/kernels/experimental/batched_rpa/flash_attention.py:58]: the q and k reshape here does not seem right....k_heads is completely missing without doing the proper adjustment for the first dimension
